### PR TITLE
chore(docs): adjust K8s documentation

### LINKED ISSUE #2094: The Best Kubernetes Setup Possible
# Subject

I would like to **reconfigure** and **document** an already documented mail server use case.

## Description

The current documentation on K8s (Kubenetes) is okay, and I was able to extract essential parts off of it. However, it lacked a bit of quality from my POV. I would like to re-do the docs on K8s. I am currently transitioning to [K3s](https://k3s.io/) and **I want to get the, especially security-wise, best "deployment"** (this is a Kubernetes-native term, but I'm referring to Services, etc. too) **possible**. I'm using Kustomize for the whole setup, and I'd document this too once finished.

I'm linking the maintainers that I know are using K8s to get some help on the topic: @radicand 
I'm also trying to get some feedback from the currently most active maintainers: @casperklein @wernerfred @polarathene
And, because in this case, more are better, I'll also link @NorseGaud, @fbartels and @reneploetz

## Manifests

I'm posting my current manifest files, that is, the deployment and the service here. I'd like to get feedback on them. Moreover, I have some questions down below.

The deployment goes first. There are a lot of file mounts that may not all be necessary, but I will explain them below.

``` YAML
---
apiVersion: apps/v1
kind: Deployment

metadata:
  name: mailserver

  annotations:
    ignore-check.kube-linter.io/run-as-non-root: >-
      The mail server needs to run as root
    ignore-check.kube-linter.io/privileged-ports: >-
      The mail server needs privilegdes ports
    ignore-check.kube-linter.io/no-read-only-root-fs: >-
      There are too many files written to make The
      root FS read-only

spec:
  replicas: 1
  selector:
    matchLabels:
      app: mailserver

  template:
    metadata:
      labels:
        app: mailserver

      annotations:
        container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default

    spec:
      hostname: mailserver
      containers:
        - name: mailserver
          image: ghcr.io/docker-mailserver/docker-mailserver:10.0.0
          imagePullPolicy: IfNotPresent

          securityContext:
            allowPrivilegeEscalation: false
            readOnlyRootFilesystem: false
            runAsUser: 0
            runAsGroup: 0
            runAsNonRoot: false
            privileged: false
            capabilities:
              add:
                # file permission capabilities
                - CHOWN
                - FOWNER
                - MKNOD
                - SETGID
                - SETUID
                - DAC_OVERRIDE
                # miscellaneous  capabilities
                - SYS_CHROOT
                - NET_BIND_SERVICE
                - KILL
              drop: [ALL]
            seccompProfile:
              type: RuntimeDefault

          resources:
            limits:
              memory: 4Gi
              cpu: 1500m
            requests:
              memory: 2Gi
              cpu: 600m

          volumeMounts:
            # OpenDKIM
            - name: opendkim-files
              subPath: mail_com.private
              mountPath: /tmp/docker-mailserver/opendkim/keys/domain1.tld/mail.private
              readOnly: true
            - name: opendkim-files
              subPath: mail_com.txt
              mountPath: /tmp/docker-mailserver/opendkim/keys/domain1.tld/mail.txt
              readOnly: true
            - name: opendkim-files
              subPath: mail_de.private
              mountPath: /tmp/docker-mailserver/opendkim/keys/domain2.tld/mail.private
              readOnly: true
            - name: opendkim-files
              subPath: mail_de.txt
              mountPath: /tmp/docker-mailserver/opendkim/keys/domain2.tld/mail.txt
              readOnly: true
            - name: opendkim-files
              subPath: KeyTable
              mountPath: /tmp/docker-mailserver/opendkim/KeyTable
              readOnly: true
            - name: opendkim-files
              subPath: SigningTable
              mountPath: /tmp/docker-mailserver/opendkim/SigningTable
              readOnly: true
            - name: opendkim-files
              subPath: TrustedHosts
              mountPath: /tmp/docker-mailserver/opendkim/TrustedHosts
              readOnly: true

            # other
            - name: files
              subPath: dh_4096.pem
              mountPath: /etc/postfix/dh_4096.pem
              readOnly: true
            - name: files
              subPath: hostname
              mountPath: /etc/hostname
              readOnly: true
            - name: files
              subPath: hosts
              mountPath: /etc/hosts
              readOnly: true

            # Postfix
            - name: files
              subPath: main.cf
              mountPath: /etc/postfix/main.cf
              readOnly: true
            - name: files
              subPath: postfix-accounts.cf
              mountPath: /tmp/docker-mailserver/postfix-accounts.cf
              readOnly: true
            - name: files
              subPath: postfix-virtual.cf
              mountPath: /tmp/docker-mailserver/postfix-virtual.cf
              readOnly: true

            # scrips
            - name: files
              subPath: setup-stack.sh
              mountPath: /usr/local/bin/setup-stack.sh
              readOnly: true
            - name: files
              subPath: user-patches.sh
              mountPath: /tmp/docker-mailserver/user-patches.sh
              readOnly: true

            # spam
            - name: files
              subPath: 50-user
              mountPath: /etc/amavis/conf.d/50-user
              readOnly: true
            - name: files
              subPath: header_checks.pcre
              mountPath: /etc/postfix/maps/header_checks.pcre
              readOnly: true
            - name: files
              subPath: postscreen.conf
              mountPath: /etc/fail2ban/filter.d/postscreen.conf
              readOnly: true
            - name: files
              subPath: user-jail.local
              mountPath: /etc/fail2ban/jail.d/user-jail.local
              readOnly: true

            # PVCs
            - name: data
              mountPath: /var/mail
              subPath: data
              readOnly: false
            - name: data
              mountPath: /var/mail-state
              subPath: state
              readOnly: false
            - name: data
              mountPath: /var/log/mail
              subPath: log
              readOnly: false

            # time
            - name: localtime
              mountPath: /etc/localtime
              readOnly: true
            - name: timezone
              mountPath: /etc/timezone
              readOnly: true

            # certificates
            - name: certificates-rsa
              mountPath: /secrets/ssl/rsa/
              readOnly: true
            - name: certificates-ecdsa
              mountPath: /secrets/ssl/ecdsa/
              readOnly: true

            # other
            - name: tmp-files
              mountPath: /tmp
              readOnly: false

          ports:
            - name: transfer
              containerPort: 25
              protocol: TCP
            - name: esmtp-implicit
              containerPort: 465
              protocol: TCP
            - name: esmtp-explicit
              containerPort: 587
            - name: imap-implicit
              containerPort: 993
              protocol: TCP

          envFrom:
            - configMapRef:
                name: mailserver.environment

      restartPolicy: Always

      volumes:
        # configuration files
        - name: files
          configMap:
            name: mailserver.other.files
        - name: opendkim-files
          configMap:
            name: mailserver.opendkim.files

        # PVCs
        - name: data
          persistentVolumeClaim:
            claimName: data

        # time
        - name: timezone
          hostPath:
            path: /etc/timezone
            type: File
        - name: localtime
          hostPath:
            path: /etc/localtime
            type: File

        # certificates
        - name: certificates-rsa
          secret:
            secretName: mail-tls-certificate-rsa
            items:
              - key: tls.key
                path: tls.key
              - key: tls.crt
                path: tls.crt
        - name: certificates-ecdsa
          secret:
            secretName: mail-tls-certificate-ecdsa
            items:
              - key: tls.key
                path: tls.key
              - key: tls.crt
                path: tls.crt

        # other
        - name: tmp-files
          emptyDir: {}

```

The service is more or less straight-forward:

``` YAML
---
apiVersion: v1
kind: Service

metadata:
  name: mailserver
  labels:
    app: mailserver

spec:
  type: LoadBalancer
  externalTrafficPolicy: Local

  selector:
    app: mailserver

  ports:
    # Transfer
    - name: transfer
      port: 25
      targetPort: transfer
      protocol: TCP
    # ESMTP with implicit TLS
    - name: esmtp-implicit
      port: 465
      targetPort: esmtp-implicit
      protocol: TCP
    # ESMTP with explicit TLS (STARTTLS)
    - name: esmtp-explicit
      port: 587
      targetPort: esmtp-explicit
      protocol: TCP
    # IMAPS with implicit TLS
    - name: imap-implicit
      port: 993
      targetPort: imap-implicit
      protocol: TCP

```

## (Possible) Improvements

The goal is simple, achieving it non-trivial:  _Define a proper ´securityContext` in the deployment_. This is, at least to me, critical in achieving the best security possible.

Especially getting

1. `readOnlyRootFilesystem:` set to `true`
2. `runAsUser:` to not be 0
3. `runAsGroup:` to not be 0
4. `runAsNonRoot:` set to `true`
5. `capabilities.drop` set to `[ALL]` and only add the caps that we really need with `capabilities.add`

This will require

1. All files that need to be adjusted on run-time to be mounted externally
2. Have a user that can manage the server as if root but without being root
3. Have a group that can manage the server as if root but without being root
4. If 2. and 3. are fulfilled, this becomes true automatically
5. Just a matter of knowing which capabilities we _actually_ need...

Feedback very welcome! I will test the deployment tomorrow at the earliest because I've got a lot to do ATM, but I will report back. In the meantime, I'd be happy to hear your ideas :D

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

This PR updates the Kubernetes documentation, making it more straight-forward, easier to understand and more structured.

Closes #2094

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] This change **is** a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/content/config/advanced/kubernetes.md [hunk 1]
```diff
@@ -460,79 +432,18 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
...
     ```
 
-**Downsides**
-
-- Not possible to access mailserver via inner cluster Kubernetes DNS, as PROXY protocol is required for incoming connections.
-
-## Let's Encrypt Certificates
+With this approach,
 
-[Kube-Lego][kube-lego] may be used for a role of Let's Encrypt client. It works with Kubernetes [Ingress Resources][k8s-network-ingress] and automatically issues/manages certificates/keys for exposed services via Ingresses.
```
Comment (polarathene) [inline docs/content/config/advanced/kubernetes.md:469]: Did this section lose relevance?

File: docs/content/config/advanced/kubernetes.md [hunk 2]
```diff
@@ -460,79 +432,18 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
...
-      - name: tls
-        mountPath: /etc/ssl/mailserver
-        readOnly: true
-    # ...
-    volumes:
-      - name: tls
-        secret:
-          secretName: mailserver.tls
-    ```
+- is is not possible to access the mail server via cluster-DNS, as the PROXY protocol is required for incoming connections
```
Comment (polarathene) [inline docs/content/config/advanced/kubernetes.md:437]:
```suggestion
- it is not possible to access the mail server via cluster-DNS, as the PROXY protocol is required for incoming connections
```

**Review by polarathene — CHANGES_REQUESTED** [review]
- [Current live docs k8s page](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/kubernetes/)
- [Current PR docs k8s page](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/)

---

Primarily minor corrections or phrasing suggestions.

- I would like to know why the last section was dropped. Other parts have been removed with no explanation either, such as the admonitions with notes/warnings in the current live docs.
- Any reason for dropping the usage of example admonitions, especially collapsible ones? Only one remains at the end.
- I understand this is WIP, some content needs to be broken apart into smaller paragraphs.
- I would like to have feedback from another k8s user/contributor, but presently the current docs seem more approachable.
- I'm not able to review the yaml configs, I trust you've done well with that part :)

File: docs/content/config/advanced/kubernetes.md [hunk 3]
```diff
@@ -2,392 +2,363 @@
...
-        app: mailserver
-      ports:
-        - name: smtp
-          port: 25
-          targetPort: smtp
+    For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMaps` and use them the same way.
+
+## Exposing your Mail Server to the Outside World
+
+The more difficult part with K8s is to expose deployed mailserver to outside world. K8s provides multiple ways for doing that; each has its downsides and complexity. The major problem with exposing the mail server to outside world in K8s is to [preserve the real client IP][k8s-service-source-ip]. The real client IP is required by the mail server for performing IP-based SPF checks and spam checks. If you do not require SPF checks for incoming mails, you may disable them in your [Postfix configuration][docs-postfix] by dropping the line stat states `check_policy_service unix:private/policyd-spf`.
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:278]: Yes :D

File: docs/content/config/advanced/kubernetes.md [hunk 4]
```diff
@@ -2,392 +2,363 @@
...
+        #  ...
+```
 
-**Downsides**
+With this approach,
 
-- Not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at.
-- Every Port within the Container is exposed on the Host side, regardless of what the `ports` section in the Configuration defines. 
+- it is not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at
+- every Port within the Container is exposed on the Host side
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:357]: No, you should read the sentences as

"With this approach, it is not possible to access mailserver via other cluster Nodes, only via the one mailserver deployed at"
"With this approach, every Port within the Container is exposed on the Host side"

> Is this a formatting preference to use lower case list items instead of capitalizing the first letter of a sentence? Likewise with the removal of full stops at the end.

I removed the full stops by accident.

> The common subheading/subsection **Downsides** is replaced with a generic and less consistent message "This approach" or similar, even though the focus is on clearly identifying the downsides to an approach?

Using `**` does not indicate a subheading in Markdown (Markdownlint will even report this). I think that this section clarifies the consequences of using the approach, which are not always downsides.

File: docs/content/config/advanced/kubernetes.md [hunk 5]
```diff
@@ -2,392 +2,363 @@
...
+              containerPort: 587
+              hostPort: 587
+            - name: imap-secure
+              containerPort: 993
+              hostPort: 993
+        #  ...
+```
 
-**Downsides**
+With this approach,
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:354]: I feel like, with the idea I had of making the lists "sentences", I feel like the ":" does not fit here.

File: docs/content/config/advanced/kubernetes.md (hunk 1 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:469]: See my comment down below. Not sure why one would use Kube-Lego, _which is deprecated btw_, instead of certManager.

**Comment by georglauterbach** [comment]
> I would like to know why the last section was dropped. Other parts have been removed with no explanation either, such as the admonitions with notes/warnings in the current live docs.

I removed the last section because on the one hand, we already have a section about certificates, on the other hand, if you want to have a dedicated section about certificates, why just write about Let'sEncrypt? I feel like with K8s, there is more to it, especially when you consider [certManager](https://cert-manager.io/docs/).

> Any reason for dropping the usage of example admonitions, especially collapsible ones? Only one remains at the end.

I found it odd that you'd need to unfold them, because this is the content everyone is there for...

> some content needs to be broken apart into smaller paragraphs.

Why do you figure? I must say I disagree.

File: docs/content/config/advanced/kubernetes.md [hunk 6]
```diff
@@ -2,392 +2,363 @@
...
-**Downsides**
+This approach
 
-- __Real client IP is not preserved__, so SPF check of incoming mail will fail.
+- does not preserve the real client IP, so SPF check of incoming mail will fail
 
 ### Bind to concrete Node and use host network
 
-The simplest way to preserve real client IP is to use `hostPort` and `hostNetwork: true` in the mailserver [Pod][k8s-workload-pod]. This comes in price of availability: you can talk to mailserver from outside world only via IPs of [Node][k8s-nodes] where mailserver is deployed.
+One way to preserve the real client IP is to use `hostPort` and `hostNetwork: true`. This comes in price of availability: you can talk to the mail server from outside world only via IPs of [Node][k8s-nodes] where mailserver is deployed.
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:324]: Just an inconsistency on my side. Will correct.

**Comment by georglauterbach** [comment]
> I would like to have feedback from another k8s user/contributor, but presently the current docs seem more approachable.

I think that with the current docs, before you're able to work with it, you'd need to expand 4 admonitions, and you do not have a clear structure. But this is just my PoV.

**Comment by polarathene** [comment]
> you'd need to expand 4 admonitions

I am not familiar with the yaml configs being shared, if all the snippets are relevant to the user you can collapse under 1 admonition for example with each clearly labeled, or if this information will always be relevant to the reader of this page, avoid any collapsing.

When there is optional configs, it often made more sense imo to reduce visual noise by collapsing it by default, a reader would scan through the document or ToC sidebar and get the information they want.

I can understand if that workflow differs for others. I personally prefer heavy pages to not be as overwhelming to someone seeking guidance by using progressive disclosure UX (_a few extra clicks at most vs overloading someone with everything all at once which usually isn't as well received, especially if only a portion is actually relevant to focus on_).

---

If the intention is to copy these configs as files, we can "import" them into the docs, they'll be rendered as they do currently, but we can additionally link to them on github to copy/paste from a clone of the repo. I believe I have instructions for that on my WIP documentation of the doc features (progress stalled due to other priorities). Would that be helpful?

---

> you do not have a clear structure

At least with the collapsed snippets, for someone familiar with k8s which seems to be the expectation here, the structure of them seemed clear that they were k8s related config and the names matched the `kind` metadata?

Your updates are helpful with the added context between them, but I think we could improve that a little, perhaps with some subheadings and not collapsing the description for each part into a single hefty paragraph? They're minor improvements compared to the value of your contribution, but they help make the content more easier to follow :sweat_smile: 

---

I removed the last section because on the one hand, we already have a section about certificates, on the other hand, if you want to have a dedicated section about certificates, why just write about Let'sEncrypt?

I don't mind it being removed, so long as we're not losing any value it may provide. If that information isn't of value or is already covered elsewhere no worries, otherwise we should perhaps migrate it somewhere more appropriate.

LetsEncrypt is popular enough that I don't mind it if it's helpful/relevant (we don't cover k8s anywhere else but this page right?). Removing it because it's only covering LetsEncrypt isn't valid, if anyone else wants to contribute alternatives that'd be welcomed, but it shouldn't be mandated if the information would be helpful.

Do you think it would be better suited and of value to the TLS docs page?

---

> I found it odd that you'd need to unfold them, because this is the content everyone is there for...

I can't really comment for this docs page as I am not familiar with what and how much of the information presented is relevant to all viewers on that page.

As stated earlier, there is value in collapsible admonitions when appropriate imo.

Regular admonitions still help with the layout of the page content and conveying the info at a glance. I would personally prefer using them but unless anyone is going to weigh in, either way is fine. The information documented is more important, presentation is complimentary to that.

---

> Why do you figure? I must say I disagree.

I find:

> A `Service` is required for getting the traffic to the pod itself. The service is somewhat crucial. Its configuration determines whether the original IP from the sender will be kept. [More about this further down below](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/#exposing-your-mail-server-to-the-outside-world). The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most K8s users will only want to have one instance anyway, and users that need high availability know how to do it anyways. You will want to have your load-balancer give this service an external, routable IP address.

to be worse UX vs:


> ### Service
> 
> A `Service` is required for getting the traffic to the pod itself. The service is somewhat crucial. Its configuration determines whether the original IP from the sender will be kept. [More about this further down below](https://pullrequest-2111--dms-doc-previews.netlify.app/config/advanced/kubernetes/#exposing-your-mail-server-to-the-outside-world).
> 
> The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most K8s users will only want to have one instance anyway, and users that need high availability know how to do it anyways.
> 
> You will want to have your load-balancer give this service an external, routable IP address.

which is easier to digest, especially in the context of the whole document.

There's already some verbosity from the writing style we could trim, but just giving the content some more breathing space really helps (_at least for me it's heavily appreciated, imagine my verbose responses compacted into just a few paragraphs.._).

If I have some k8s familiarity and have a setup already or looking back at these docs to reference something, is it possible that I might want to quickly lookup one of these manifests as reference? With the subheading it'll be added to the ToC sidebar under the mainfests heading level, and it'll be easier to scroll through the document and locate (_if not using the ToC_).

**Comment by georglauterbach** [comment]
About the TLS / certificates topic: Kube-Lego is deprecated, and should not be used in new deployments anymore. You'd want to use CertManager. However, there are many possibilities. That's why I'd leave this section (or mention it in the TLS section seperately).

I have added more subsections to structure the document a bit more. I would however very much try to avoid putting the configuration files in admonitions that one has to unfold. With recent versions of MkDocs (which we are using), there is a dedicated copy button in the top right hand corner one can use to copy the code as a whole :)

I have also split the description for the service.

File: docs/content/config/advanced/kubernetes.md [hunk 7]
```diff
@@ -460,79 +444,19 @@ Then, configure both [Postfix][docs-postfix] and [Dovecot][docs-dovecot] to expe
...
-    # ...
-    volumes:
-      - name: tls
-        secret:
-          secretName: mailserver.tls
-    ```
+- it is not possible to access the mailserver via cluster-DNS, as the PROXY protocol is required for incoming connections.
 
+[Kustomize]: https://kustomize.io/
+[TLS]: ../security/ssl.md
```
Comment (polarathene) [inline docs/content/config/advanced/kubernetes.md:452]: Convention for doc links is using the `docs-` prefix. I often keep it all lowercase, the idea with the prefix is that it'd be easy to parse docs with tooling or CLI utilities for use cases like identifying/checking internal doc links :man_shrugging: 

Maybe it's over thinking, and isn't necessary.

I wanted to add it to our CI at one point so our docs wouldn't miss changes breaking links. The Mkdocs build --strict feature generally works, but I do recall it having some gotchas such as not catching invalid anchor (`#)` links to jump to specific sections of our docs.

File: docs/content/config/advanced/kubernetes.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (polarathene) [reply docs/content/config/advanced/kubernetes.md]: > No, you should read the sentences as

Yeah, I still do when presented with a list. Personal bias of mine perhaps that the list items still are formatted like new sentences starting with capital letters. I don't mean to be too fussy about it :)

It's a little weird at the end of the doc with a single list item. Would a numbered list work better? Up to you, leave it if you think it's fine as-is.

> Using ** does not indicate a subheading in Markdown (Markdownlint will even report this).

Yes, the issue with actual heading is it's invalid for them to all have the same heading. At the time I didn't know how to resolve that, but I'm later learned of syntax to provide unique heading ids if we wanted the ToC and links to headings to work correctly.

> which are not always downsides.

Fair enough :+1:

File: docs/content/config/advanced/kubernetes.md (hunk 7 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md]: Done with af58f352188a16c0e12bab0ec033e7eb56838531 :)

**Comment by georglauterbach** [comment]
I will test this ASAP, but I have to figure out how solve a problem with [Cilium](https://cilium.io/) in my cluster first. I will mark this as ready for review as soon as I know my mail server setup works.

File: docs/content/config/advanced/kubernetes.md [hunk 8]
```diff
@@ -448,8 +448,8 @@ With this approach,
 
 - it is not possible to access the mailserver via cluster-DNS, as the PROXY protocol is required for incoming connections.
 
-[Kustomize]: https://kustomize.io/
-[TLS]: ../security/ssl.md
+[docs-kustomize]: https://kustomize.io/
```
Comment (polarathene) [inline docs/content/config/advanced/kubernetes.md:451]: I think there was a misunderstanding here, `kustomize` is an external link, not internal docs.

Usually that distinction is clear as I placed all the internal `docs-` links above the others. For example below is `dockerhub-haproxy`. You can continue to use `kustomize`, but `docs-` is misleading here :sweat_smile:

File: docs/content/config/advanced/kubernetes.md [hunk 9]
```diff
@@ -2,392 +2,375 @@
...
+  MOVE_SPAM_TO_JUNK: '1'
+  ENABLE_UPDATE_CHECK: '1'
+  ENABLE_SPAMASSASSIN: '1'
+  SUPERVISOR_LOGLEVEL: warn
+  SPAMASSASSIN_SPAM_TO_INBOX: '1'
+```
 
-!!! warning
-    Any sensitive data (keys, etc) should be deployed via [Secrets][k8s-config-secret]. Other configuration just fits well into [ConfigMaps][k8s-config-pod].
+We can also make use of user-provided configuration files, e.g. `user-patches.sh`, `postfix-accounts.cf` and more, to adjust `docker-mailserver` to our likings. We encourage you to have a look at [Kustomize][docs-kustomize] for creating `ConfigMap`s from multiple files, but for now, we will provide a simple, hand-written example. This example is absolutely minimal and only goes to show what can be done.
```
Comment (polarathene) [inline docs/content/config/advanced/kubernetes.md:56]:
```suggestion
We can also make use of user-provided configuration files, e.g. `user-patches.sh`, `postfix-accounts.cf` and more, to adjust `docker-mailserver` to our likings. We encourage you to have a look at [Kustomize][kustomize] for creating `ConfigMap`s from multiple files, but for now, we will provide a simple, hand-written example. This example is absolutely minimal and only goes to show what can be done.
```

Or if you prefer the `[Kustomize]` alone is fine with me too.

**Review by polarathene — CHANGES_REQUESTED** [review]

File: docs/content/config/advanced/kubernetes.md (hunk 8 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:451]: I will change this, but I need to adjust the link above so I cannot directly commit your suggested changes.

File: docs/content/config/advanced/kubernetes.md (hunk 9 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:56]: I just now saw this, alright, I can commit your changes :D

**Review by polarathene — APPROVED** [review]
Ideally someone with k8s knowledge can also chime in, but the rest seems good :+1:

**Comment by georglauterbach** [comment]
@polarathene There will be Orr commits because I will remove small parts (I overlooked a few things when copy-wasting my setup). I will reach back to you and the others when this is ready again :)

**Comment by radicand** [comment]
I'll try to review when I get some time

**Comment by georglauterbach** [comment]
Ready for review.

**Review by polarathene — DISMISSED** [review]

**Comment by georglauterbach** [comment]
@polarathene Sorry for dismissing your review again. Because of #2113 I realised I forgot to update the caps list :)

**Review by polarathene — DISMISSED** [review]

**Comment by georglauterbach** [comment]
@radicand Have you had time to get a look at the updates?:)

**Comment by radicand** [comment]
> @radicand Have you had time to get a look at the updates?:) If you don't have the time, I think it's not a big deal, I'd just need to know :D

Starting now...

File: docs/content/config/advanced/kubernetes.md [hunk 10]
```diff
@@ -2,392 +2,374 @@
...
+metadata:
+  name: data
+
+spec:
+  storageClassName: local-path
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:90]: I'd suggest we either change this default to something larger, or make it explicit (as in providing guidance) on setting it. While PVCs can be resized, it is not _always easy_ to do based on the storage backend, and running out of space prematurely sours the experience for sure.

File: docs/content/config/advanced/kubernetes.md [hunk 11]
```diff
@@ -2,392 +2,374 @@
...
-    metadata:
-      name: mailserver
-    # ...
-        volumeMounts:
-          - name: config
-            subPath: postfix-main.cf
-            mountPath: /tmp/docker-mailserver/postfix-main.cf
-            readOnly: true
-    ```
+With this setup, you want your load-balancer (e.g. MetalLB) to give this service an external, especially routable IP address.
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:99]: You do not need MetalLB to give you an external address - an internal address that you DNAT/port forward will do (this is how I run it actually). Since there is a later section talking about MetalLB anyway, I'd strike the line here and just reference that real client IP is discussed below.

File: docs/content/config/advanced/kubernetes.md [hunk 12]
```diff
@@ -2,392 +2,374 @@
...
+      targetPort: imap-implicit
+      protocol: TCP
+
+```
 
-The simplest way is to expose mailserver as a [Service][k8s-network-service] with [external IPs][k8s-network-external-ip].
+### Deployments
 
-!!! example
+Last but not least, the `Deployment` becomes the most complex component. The deployment brings maximal security measures without compromising on ease of use.
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:144]: I find this a somewhat confusing sentence -- I'd suggest we just mention what the purpose of the Deployment object is rather than talking about security (which you can do, but a deployment's purpose is not to enforce security). Something like "Lastly, the Deployment is the most complex component. It instructs Kubernetes how to run the docker-mailserver container and how to apply your ConfigMaps and persisted storage. Additionally, we can set options to enforce runtime security here."

File: docs/content/config/advanced/kubernetes.md [hunk 13]
```diff
@@ -2,392 +2,374 @@
...
+            seccompProfile:
+              type: RuntimeDefault
+
+          resources:
+            limits:
+              memory: 4Gi
+              cpu: 1500m
+            requests:
+              memory: 2Gi
+              cpu: 600m
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:217]: We might want to add comments that these values can/should be tuned based on the scale of the implementation. I run the below for 2 (real) users, ClamAV off, and ~400 emails a day. Spam sucks.
```
            requests:
              memory: "500Mi"
              cpu: "750m"
            limits:
              memory: "1500Mi"
              cpu: "900m"
```

I mention this because K8s will start refusing to schedule pods if you exceed your memory/cpu budgets, and you don't _always_ need the suggested amounts. Users should be aware that they can tune this.

File: docs/content/config/advanced/kubernetes.md [hunk 14]
```diff
@@ -2,392 +2,374 @@
...
+              mountPath: /tmp
+              readOnly: false
+
+          ports:
+            - name: transfer
+              containerPort: 25
+              protocol: TCP
+            - name: esmtp-implicit
+              containerPort: 465
+              protocol: TCP
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:250]: This may be debatable, but 465 isn't a supported standard from what I'm aware, and nothing requires it (famous last words). Should we be good citizens and not advertise opening it up by removing it from here and the service?

File: docs/content/config/advanced/kubernetes.md [hunk 15]
```diff
@@ -2,392 +2,374 @@
...
+          emptyDir: {}
+```
+
+### Sensitive Data
+
+By now, the mailserver starts, but does not really work for long (or at all), because we're lacking certificates. You will need to choose yourself, which approach you'd want to go with. The [TLS][docs-tls] section provides you with an overview.
+
+!!! attention "Sensitive Data"
+
+    For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMaps` and use them the same way.
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:285]: Just pointing out a typo in removing the `s` from ConfigMaps like is done in Secrets just prior.

File: docs/content/config/advanced/kubernetes.md [hunk 16]
```diff
@@ -2,392 +2,374 @@
...
+
+## Exposing your Mailserver to the Outside World
```
Comment (radicand) [inline docs/content/config/advanced/kubernetes.md:287]: For this section, I'd suggest we start with a recommendation, which is to use MetalLB and route to the LB IP via DNAT, direct routing, etc. If MetalLB can't be used, then here is our next best suggestion, but its caveats are...

This doesn't have to be changed, but as I mentioned above, an externally routable IP is not required for SPF to work when using a LB like MetalLB. I think we're just providing an impression that users need something that is not often available (a dedicated external IP they can use for this), when it's not really the case.

**Comment by radicand** [comment]
I'm done with my review, feel free to take or leave any of my points :)

File: docs/content/config/advanced/kubernetes.md (hunk 14 above)
Comment (polarathene) [reply docs/content/config/advanced/kubernetes.md:250]: 465 for implicit TLS is valid, it has a bit of a rough history being deprecated from STMPS (_port 25 with implicit TLS_) to being designated as implicit TLS `submissions` service (_as opposed to explicit `submission` service with StartTLS on port 587_).

That change requires auth like port 587, but plain TLS socket can be used, no StartTLS negotiation required. Due to the confusion from it's history, it's support is a bit spotty when offered by services.

587 had been around for a while and established itself, so is more common to see and without any confusion. That said StartTLS has also had some issues with how some software behaved (_eg IIRC.. sending auth credentials in plain-text if StartTLS failed to negotiate a secure exchange and the server rejected the client)).

File: docs/content/config/advanced/kubernetes.md (hunk 12 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:144]: I find your sentence better too. I will copy it :)

File: docs/content/config/advanced/kubernetes.md (hunk 14 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:250]: It's funny to see this debate everywhere :D @radicand you may want to read [this](https://github.com/authelia/authelia/issues/2194), and what @polarathene said is absolutely correct :)

TL;DR 465 is actually preferred over 578

File: docs/content/config/advanced/kubernetes.md (hunk 15 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:285]: Done.

File: docs/content/config/advanced/kubernetes.md (hunk 16 above)
Comment (georglauterbach) [reply docs/content/config/advanced/kubernetes.md:287]: I will see what I can do, but I like the idea :)

**Comment by georglauterbach** [comment]
Alright. Thank you @radicand for your suggestions :D

@polarathene I have enabled auto-merge. If you think this is okay, this PR should get merged when you approve it :)

**Review by radicand — APPROVED** [review]
LGTM

**Comment by georglauterbach** [comment]
@polarathene You have already approved this PR a thousand times. Therefore, I'm going to assume you're fine with it and I'm beating you to the punch (a strange English phrase) with merging this PR so you don't have to approve it again :D

**Review by polarathene — APPROVED** [review]
